pool.cpp

```diff
@@ -0,0 +1,64 @@
+/**
+ *  Copyright 2014-2015 MongoDB, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "hphp/runtime/ext/extension.h"
+
+#include "pool.h"
+
+extern "C" {
+#include "../../../libmongoc/src/mongoc/mongoc-client.h"
+}
+
+namespace {
+	thread_local std::unordered_map<std::string, std::shared_ptr<HPHP::Pool>> s_connections;
+}
+
+namespace HPHP {
+
+std::string Pool::GetHash(const mongoc_uri_t *uri)
+{
+	return mongoc_uri_get_string(uri);
```

Comment:
I don't believe this is sufficient, since it ignores additional [connection string options](https://docs.mongodb.org/manual/reference/connection-string/#connection-string-options) provided in the `MongoDB\Driver\Manager` array argument. Perhaps the URI should be be combined with serialized array options. There is still the issue of functionally equivalent being considered different (e.g. order of keys in the options array), which no doubt requires a non-trivial amount of effort to sort out.

On a side note, I suppose this is a performance regression. We could previously re-use sockets between clients with differing, irrelevant options (e.g. default write concern, read preference), but if the pool is doling out clients to us, we have no choice but to keep unique clients per distinct "URI + option" hash. I suppose we could discard such irrelevant options from the hash and then apply them each time we request a client from the pool. That should work as long as we guarantee that the same client is never checked out of the pool concurrently.